chapters/en/chapter5/fine-tuning_parakeet.mdx

```diff
@@ -0,0 +1,361 @@
+</Tip>
+
+We use streaming mode since MLS is a large dataset. Let's cast the audio to the correct sampling rate:
+
+```python
+mls_train = mls_train.cast_column("audio", Audio(sampling_rate=16000))
+mls_test = mls_test.cast_column("audio", Audio(sampling_rate=16000))
+```
+
+## Preprocessing
+
+Now let's load the processor:
+
+```python
+from transformers import AutoProcessor
+
+processor = AutoProcessor.from_pretrained("nvidia/parakeet-ctc-1.1b")
+```
+
+Since Parakeet was trained on English, its vocabulary may not cover all Spanish characters. However, the model uses a character-based tokenizer that can handle most Latin characters. For optimal results on languages with special characters, you might need to extend the vocabulary, but for Spanish the existing tokenizer works reasonably well.
+
+### Data Collator
+
+We need a data collator to process the audio and text inputs correctly. The Parakeet processor handles feature extraction for the audio and tokenization for the text.
+
+```python
+class ParakeetDataCollator:
+    """Data collator for Parakeet CTC training."""
+
+    def __init__(self, processor):
+        self.processor = processor
+
+    def __call__(self, features):
+        texts = [f["transcript"] for f in features]
+        audios = [f["audio"]["array"] for f in features]
+
+        # Process audio features and text labels
+        inputs = self.processor(
+            audio=audios,
+            text=texts,
+            sampling_rate=self.processor.feature_extractor.sampling_rate,
+        )
+        return inputs
+```
+
+## Training (Full Fine-tuning)
+
+For full fine-tuning, we simply load the model and train all parameters:
+
+```python
+from transformers import AutoModelForCTC, TrainingArguments, Trainer
+
+model = AutoModelForCTC.from_pretrained("nvidia/parakeet-ctc-1.1b")
+
+data_collator = ParakeetDataCollator(processor)
+
+training_args = TrainingArguments(
+    output_dir="parakeet-spanish-finetuned",
+    max_steps=1000,  # Use max_steps for streaming datasets
+    per_device_train_batch_size=2,
+    gradient_accumulation_steps=4,
+    learning_rate=5e-5,
+    bf16=True,
+    logging_steps=10,
+    eval_strategy="steps",
+    eval_steps=200,
+    save_strategy="steps",
+    save_steps=200,
+    remove_unused_columns=False,
+    report_to="tensorboard",
+    push_to_hub=True,
+)
+
+trainer = Trainer(
+    model=model,
+    args=training_args,
+    train_dataset=mls_train,
+    eval_dataset=mls_test,
+    data_collator=data_collator,
+)
+
+trainer.train()
+```
+
+## Efficient Training with LoRA
+
+Fine-tuning a 1.1B parameter model can be memory-intensive. To fit this on a standard GPU, we can use Parameter-Efficient Fine-Tuning (PEFT) with LoRA:
+
+1.  **LoRA (Low-Rank Adaptation)**: Freezes most model weights and only trains small adapter layers.
+2.  **Gradient Checkpointing**: Trades compute for memory by re-computing activations during the backward pass.
+
+Let's configure the model for LoRA:
```

Comment:
Let's configure the model for LoRA with the `peft` library ([docs](https://huggingface.co/docs/peft/en/index)):